config/deploy/docker/lib/aws_sdk_helpers.rb

```diff
@@ -0,0 +1,105 @@
+    define_singleton_method(:ec2) { Aws::EC2::Client.new}
+    define_singleton_method(:cloudwatchevents) { Aws::CloudWatchEvents::Client.new }
+    define_singleton_method(:cw)  { Aws::CloudWatch::Client.new }
+    def self.cwl(profile = nil)
+      profile ||= AwsSdkHelpers::Helpers.cluster_name
+      Aws::CloudWatchLogs::Client.new(profile: profile)
+    end
+
+    define_method(:iam)              { AwsSdkHelpers::ClientMethods.iam }
+    define_method(:elbv2)            { AwsSdkHelpers::ClientMethods.elbv2 }
+    define_method(:ecr)              { AwsSdkHelpers::ClientMethods.ecr }
+    define_method(:ecs)              { AwsSdkHelpers::ClientMethods.ecs }
+    define_method(:secretsmanager)   { AwsSdkHelpers::ClientMethods.secretsmanager }
+    define_method(:autoscaling)      { AwsSdkHelpers::ClientMethods.autoscaling }
+    define_method(:ec2)              { AwsSdkHelpers::ClientMethods.ec2 }
+    define_method(:cloudwatchevents) { AwsSdkHelpers::ClientMethods.cloudwatchevents }
+    define_method(:cw)               { AwsSdkHelpers::ClientMethods.cw }
+    define_method(:cwl)              { AwsSdkHelpers::ClientMethods.cwl(
+      profile: respond_to?(:_cluster_name) ? _cluster_name : AwsSdkHelpers::Helpers.cluster_name,
+    ) }
+  end
+
+  module Helpers
+    include AwsSdkHelpers::ClientMethods
+
+    def self.cluster_name
+      ENV.fetch('CLUSTER_NAME', ENV.fetch('AWS_CLUSTER', ENV.fetch('AWS_PROFILE', ENV.fetch('AWS_VAULT', ''))))
+    end
+    def _cluster_name
+      AwsSdkHelpers::Helpers.cluster_name
+    end
+
+    def self.capacity_providers(cluster)
+      r = {}
+      capacity_provider_names = AwsSdkHelpers::ClientMethods.ecs.describe_clusters(clusters: [cluster]).clusters.first.capacity_providers
+      AwsSdkHelpers::ClientMethods.ecs.describe_capacity_providers(capacity_providers: capacity_provider_names).capacity_providers.map do |capacity_provider|
+        asg_name = capacity_provider.auto_scaling_group_provider.auto_scaling_group_arn.split('/').last
+        asg = AwsSdkHelpers::ClientMethods.autoscaling.describe_auto_scaling_groups(auto_scaling_group_names: [ asg_name ]).auto_scaling_groups.first
+
+        launch_template_id = asg.mixed_instances_policy.launch_template.launch_template_specification.launch_template_id
+        launch_template_versions = AwsSdkHelpers::ClientMethods.ec2.describe_launch_template_versions(launch_template_id: launch_template_id)
+
+        ami_id = launch_template_versions.launch_template_versions[0].launch_template_data.image_id
+
+        r[capacity_provider.name] = {
+          name: capacity_provider.name,
+          ami_id: ami_id
+        }
+      end
+      return r
+    end
+    def _capacity_providers(cluster = nil)
+      cluster ||= _cluster_name
+      @capacity_providers ||= AwsSdkHelpers::Helpers.capacity_providers(cluster)
+    end
+
+    def self.default_placement_constraints(ami_id:)
+      [
+        {
+          expression: "attribute:ecs.ami-id == #{ami_id}",
```

Comment:
What's the plan for handling the situation right after we publish a new launch template? I think we're fine for everything already running when that moment happens, but I think scheduled tasks, new service deployments, and the occasional container crash will find the new containers have no place to go.

I think this constraint needs to be more complex (or the code above) so that we only constrain if there's actually ec2 instances running that could fulfill the request and that have capacity.

Reply:
Added logic to query ec2.describe_instances for capacity provider instances w/matching ami.

Reply:
@yakloinsteak @eanders We need to keep in mind here that, unless I'm misinterpreting the code, we're looking at these constraints only being guaranteed to be re-evaluated during deployment. I think that's fine, and the process would look something like:

- update ami
- update capacity_provider config to use new ami_id
- application now has a constraint on the old ami_id, and since those instances stick around while they're being used, we're in a safe place here
- deploy application, which will update the constraints to use the new ami_id (unless no instances w/that ami are available yet)
- now the application is only able to spin up new things on the new ami, and things start moving over

This isn't like, a problem, I just wanted to call out that the constraints aren't necessarily going to be re-evaluated every time, say, a new cron task is spun up.